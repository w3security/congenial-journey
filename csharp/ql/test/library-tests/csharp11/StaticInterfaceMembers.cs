public interface INumber<T> where T : INumber<T>
{
    static abstract T operator ++(T other);

    static virtual T operator --(T other) => other;

    static abstract T Add(T left, T right);

    static virtual T Subtract(T left, T right) => left;

    static T Zero() => default(T);
}

public class Complex : INumber<Complex>
{
    public double Real { get; private set; } = 0.0;
    public double Imaginary { get; private set; } = 0.0;

    public Complex() { }

    public static Complex Zero() => new Complex();

    public static Complex operator ++(Complex other) =>
        new Complex { Real = other.Real + 1.0, Imaginary = other.Imaginary };

    public static Complex operator --(Complex other) =>
        new Complex { Real = other.Real - 1.0, Imaginary = other.Imaginary };

    public static Complex Add(Complex left, Complex right) =>
        new Complex { Real = left.Real + right.Real, Imaginary = left.Imaginary + right.Imaginary };

    public static Complex Subtract(Complex left, Complex right) =>
        new Complex { Real = left.Real - right.Real, Imaginary = left.Imaginary - right.Imaginary };
}