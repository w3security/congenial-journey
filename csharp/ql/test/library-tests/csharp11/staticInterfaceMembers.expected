interfacemembers
| INumber<> | StaticInterfaceMembers.cs:3:32:3:33 | ++ | abstract |
| INumber<> | StaticInterfaceMembers.cs:3:32:3:33 | ++ | public |
| INumber<> | StaticInterfaceMembers.cs:3:32:3:33 | ++ | static |
| INumber<> | StaticInterfaceMembers.cs:5:31:5:32 | -- | public |
| INumber<> | StaticInterfaceMembers.cs:5:31:5:32 | -- | static |
| INumber<> | StaticInterfaceMembers.cs:5:31:5:32 | -- | virtual |
| INumber<> | StaticInterfaceMembers.cs:7:23:7:25 | Add | abstract |
| INumber<> | StaticInterfaceMembers.cs:7:23:7:25 | Add | public |
| INumber<> | StaticInterfaceMembers.cs:7:23:7:25 | Add | static |
| INumber<> | StaticInterfaceMembers.cs:9:22:9:29 | Subtract | public |
| INumber<> | StaticInterfaceMembers.cs:9:22:9:29 | Subtract | static |
| INumber<> | StaticInterfaceMembers.cs:9:22:9:29 | Subtract | virtual |
| INumber<> | StaticInterfaceMembers.cs:11:14:11:17 | Zero | public |
| INumber<> | StaticInterfaceMembers.cs:11:14:11:17 | Zero | static |
implements
| StaticInterfaceMembers.cs:23:36:23:37 | ++ | StaticInterfaceMembers.cs:3:32:3:33 | ++ |
| StaticInterfaceMembers.cs:26:36:26:37 | -- | StaticInterfaceMembers.cs:5:31:5:32 | -- |
| StaticInterfaceMembers.cs:29:27:29:29 | Add | StaticInterfaceMembers.cs:7:23:7:25 | Add |
| StaticInterfaceMembers.cs:32:27:32:34 | Subtract | StaticInterfaceMembers.cs:9:22:9:29 | Subtract |
